### PR TITLE
[dynamo][numpy] Handle return value being numpy ndarray

### PR DESCRIPTION
**Pull request by larryliu0820**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #99560
* #95849

On top of #95849 this PR is trying to handle the special case when dealing with numpy.

Consider the following example:

```
def f(x: torch.Tensor) -> np.ndarray:
	a = x.numpy()
	return a.T
```
In previous PR this will error out because we translate `a.T` to be a method call on `torch_np.ndarray.T` which is also a `torch_np.ndarray`.

This PR handles this case, by conditionally converting a `torch_np.ndarray` to `np.ndarray` before returning, to match the original behavior.

The compiled version will be:

```
def f(x):
    ___tmp_0 = __compiled_fn_0(x)
    if isinstance(___tmp_0, torch_np.ndarray):
        return ___tmp_0.tensor.numpy()
    else:
        return ___tmp_0
```

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: torch/_dynamo/variables/tensor.py [hunk 1]
```diff
@@ -720,6 +722,55 @@ def call_method(
     ) -> "VariableTracker":
         unimplemented(f"numpy_ndarray.{name}()")
 
+    @staticmethod
+    def reconstruct_ndarray_before_return(codegen, output_graph) -> List[Instruction]:
```
Comment (jansel) [inline torch/_dynamo/variables/tensor.py:726]: You might be able to simplify this by defining:
```
def to_numpy_helper(___tmp_0):
    if isinstance(___tmp_0, torch_np.ndarray):
        return ___tmp_0.tensor.numpy()
    else:
        return ___tmp_0
```

Then calling it similar to:
https://github.com/pytorch/pytorch/blob/9244264f469036ddd8868cece96f640bdbd2ccfc/torch/_dynamo/side_effects.py#L342

File: torch/_dynamo/variables/tensor.py [hunk 2]
```diff
@@ -720,6 +722,55 @@ def call_method(
...
+            return []
+        from torch._dynamo.bytecode_transformation import create_instruction
+
+        # b.tensor.numpy()
+        var = output_graph.new_var()
+        from torch_np import ndarray
+
+        check = [
+            codegen.create_store(var),
+            codegen.create_load_global("isinstance", False, add=True),
```
Comment (jansel) [inline torch/_dynamo/variables/tensor.py:747]: The user could theoretically shadow this name in their global scope.

**Review by jansel — APPROVED** [review]

File: torch/_dynamo/variables/tensor.py (hunk 2 above)
Comment (yanboliang) [reply torch/_dynamo/variables/tensor.py:747]: Yea, you can use this helper function to mitigate:
https://github.com/pytorch/pytorch/blob/ee5f09ab802a04cc966829230c236e9132041ab7/torch/_dynamo/codegen.py#L291

but a better way is to replace these bytecode by ```to_numpy_helper```.

**Review by yanboliang — APPROVED** [review]

**Pull request closed without merging**